### DISCUSSION TITLE
How to disable WARNING! There are options you set that were not used!

### DISCUSSION THREAD

**Discussion by andrewlhicks** [firedrakeproject/firedrake#3770] [Firedrake support] [answered]
I use Firedrake by running a script that takes its own command line arguments. I think PETSc doesn't like this very much. Every time I run my code, I get the following warnings:
```
WARNING! There are options you set that were not used!
WARNING! could be spelling mistake, etc!
```
As mentioned [here](https://community.freefem.org/t/suppress-options-left-when-using-petsc/2702), these warnings can be suppressed by passing `-options_left no` into the command line.

But quite frankly, I don't want to do that every single time I run my code. Is there possibly a line of code I can simply add into my script which has the same effect?

**Comment by pbrubeck** [comment]
Why are you getting this warning? You are probably getting it from the combination of `argparse` to read options in your script and PETSc. The cleanest way to avoid the warning is to read options with PETSc exclusively:
```
from firedrake import *
from firedrake.petsc import PETSc

opts = PETSc.Options()
n = opts.getInt("n", default=3)
mesh = UnitSquareMesh(n, n)
```

Note that you need to pass your options with single hyphen when you run your script:
```
python sript.py -n 10
```

**Reply by andrewlhicks** [reply]
Interesting. Actually, I am using `sys.argv` to read the options, and `getopt` to parse them.

The main problem with trying to switch to PETSc for reading the options is that some of my options have the double `--` flag instead of the single `-` flag. Can PETSc handle these too?

**Reply by pbrubeck** [reply]
Yes, it can
```
n = opts.getInt("--n", default=3)
```

**Reply by andrewlhicks** [reply]
I'm open to trying this method at some point, but rewriting my script would take some time, so I'll try some point later and give an update.

**Comment by pbrubeck** [comment]
The direct answer to your question is to pass `-options_left 0` as default to the non-PETSc options parser. In the case of `argparse` it would be something like
```
parser.add_argument('-options_left', type=int, default=0,
                    help='turn off annoying message')
```

**Reply by andrewlhicks** [reply]
Tried it on a test script:
```
from firedrake import *
import argparse

parser = argparse.ArgumentParser(description='Process some integers.')
parser.add_argument('-options_left', type=int, default=0,
                    help='turn off annoying message')
```
Unfortunately, didn't work. I also tried redefining `sys.argv` to include `-options_left 0` to no avail. Maybe there is another way?

**Reply by connorjward** [reply]
I think this doesn't work because the line `from firedrake import *` is where `sys.argv` gets passed through to initialise PETSc. Any tinkering with options would need to be done prior to Firedrake import.

**Reply by andrewlhicks** [reply, marked as answer]
Just tried it, doesn't work. However, your solution inspired me to come up with a solution of my own using `sys.argv` instead of `argparse`. Here are the contents of a test script which I've named `script.py`:
```
import sys

sys.argv.insert(-1, '-options_left')
sys.argv.insert(-1, '0')

print(sys.argv)

from firedrake import *
```
Next, I ran the command `python script.py -opt arg HELLOWORLD` as a test. The output of this command is:
```
['script.py', '-opt', 'arg', '-options_left', '0', 'HELLOWORLD']
```
The warning I was receiving before is now gone. Not only that, but this solution doesn't mess up any listargs (such as `HELLOWORLD`) that may come after the normal options/args.

**Comment by ReHoss** [comment]
Can this be handled differently, for instance with an env. variable? I think something cleaner than modifying the python source should be found.

**Comment by JHopeCollins** [comment]
You can create a petscrc file and add `-options_left 0` to that. This will apply the petsc options in the rc file every time petsc is called. You can still override them in the code or on the command line for specific instances where you want different behaviour.
https://petsc.org/main/manual/other/#the-options-database

**Reply by connorjward** [reply]
This is good to know!
